### ISSUE TITLE
rocBLAS no longer builds

### ISSUE THREAD

**Issue by pfultz2** [ROCmSoftwarePlatform/rocBLAS#664]
When building rocBLAS from 2.7 release(ie from [1ac2271e72d21fa745c7af87cd5fd102f90cba45](https://github.com/ROCmSoftwarePlatform/rocBLAS/commit/1ac2271e72d21fa745c7af87cd5fd102f90cba45)) we get an error from cmake:

```
Please run this script after at least one build of rocBLAS.
library/src/CMakeFiles/rocblas.dir/build.make:697: recipe for target 'library/src/librocblas.so.0.1' failed
```

**Comment by pfultz2** [comment]
It seems the `POST_BUILD` process is the problem [here](https://github.com/ROCmSoftwarePlatform/rocBLAS/blob/bae39ccb155f009660f9d9047d917449cee55157/library/src/CMakeLists.txt#L212). It seems `header_compilation_tests.sh` should run during testing not during a regular build of rocBLAS.

**Comment by justxi** [comment]
The script https://github.com/ROCmSoftwarePlatform/rocBLAS/blob/f1b6bb43594b6652e00c39d1df248d836b878df9/header_compilation_tests.sh#L8 has a fixed path. It depends where you have your build directory. If you have your build directory outside of the source tree, it fails on my tests.

**Comment by xdgarrido** [comment]
Rocblas is not building for ROCM 2.7 even if using install.sh. Please, try Rocm 2.7 .



The rocBLAS checked out from GitHub branch master is failing to build, this is the output:-

 

[ 11%] Building CXX object library/src/CMakeFiles/Tensile.dir/__/__/Tensile/Tools.cpp.o

clang-9: warning: -amdgpu-target argument 'gfx908' is not recognized; using gfx803 instead [-Winvalid-command-line-argument]

clang-9: warning: -amdgpu-target argument 'gfx908' is not recognized; using gfx803 instead [-Winvalid-command-line-argument]

clang-9: warning: -amdgpu-target argument 'gfx908' is not recognized; using gfx803 instead [-Winvalid-command-line-argument]

clang-9: warning: -amdgpu-target argument 'gfx908' is not recognized; using gfx803 instead [-Winvalid-command-line-argument]

clang-9: warning: -amdgpu-target argument 'gfx908' is not recognized; using gfx803 instead [-Winvalid-command-line-argument]

[ 11%] Built target rocblas-test-data

clang-9: error: unable to execute command: Killed

clang-9: error: clang frontend command failed due to signal (use -v to see invocation)

HCC clang version 9.0.0 (/data/jenkins_workspace/compute-rocm-rel-2.7/external/hcc-tot/clang 5c2570257a8bbd74eff632fbc60692ef61ef8ecb) (/data/jenkins_workspace/compute-rocm-rel-2.7/external/hcc-tot/compiler 9a9477021e6998100ff64d1360dcfe64f65cebe5) (based on HCC 2.7.19315-346267d-5c25702-9a94770 )

Target: x86_64-unknown-linux-gnu

Thread model: posix

InstalledDir: /opt/rocm/bin

clang-9: note: diagnostic msg: PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.

clang-9: note: diagnostic msg: Error generating preprocessed source(s).

library/src/CMakeFiles/Tensile.dir/build.make:62: recipe for target 'library/src/CMakeFiles/Tensile.dir/__/__/Tensile/Kernels.cpp.o' failed

make[2]: *** [library/src/CMakeFiles/Tensile.dir/__/__/Tensile/Kernels.cpp.o] Error 254

CMakeFiles/Makefile2:110: recipe for target 'library/src/CMakeFiles/Tensile.dir/all' failed

make[1]: *** [library/src/CMakeFiles/Tensile.dir/all] Error 2

Makefile:151: recipe for target 'all' failed

 

xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

This is my env !

 

apt show rocm-libs -a

 

Version: 2.7.22

Priority: optional

Section: devel

Maintainer: Advanced Micro Devices Inc.

Installed-Size: 13.3 kB

Depends: rocfft, rocrand, rocblas, hipblas, rocsparse, hipsparse, rocalution

Homepage: https://github.com/RadeonOpenCompute/ROCm

Download-Size: 784 B

APT-Sources: http://repo.radeon.com/rocm/apt/debian xenial/main amd64 Packages

Description: Radeon Open Compute (ROCm) Runtime software stack

 

diego@diego-MS-7B09:[master]~/Git/rocBLAS$  dpkg -l | grep rock

ii  rock-dkms                  2.7-22             all                rock-dkms driver in DKMS format.

diego@diego-MS-7B09:[master]~/Git/rocBLAS$ dpkg -l | grep rocm

ii  rocblas                    2.4.0.1471-rocm-re amd64              Radeon Open Compute BLAS library

ii  rocm-clang-ocl             0.4.0.35-rocm-rel- amd64              OpenCL compilation with clang compiler.

ii  rocm-cmake                 0.3.0.93-rocm-rel- amd64              rocm-cmake built using CMake

ii  rocm-dev                   2.7.22             amd64              Radeon Open Compute (ROCm) Runtime software stack

ii  rocm-device-libs           0.0.1              amd64              Radeon Open Compute - device libraries

ii  rocm-dkms                  2.7.22             amd64              Radeon Open Compute (ROCm) Runtime software stack

ii  rocm-opencl                1.2.0-2019080925   amd64              OpenCL/ROCm

ii  rocm-opencl-dev            1.2.0-2019080925   amd64              OpenCL/ROCm

ii  rocm-smi                   1.0.0-167-ge14d23e amd64              System Management Interface for ROCm

ii  rocm-utils                 2.7.22             amd64              Radeon Open Compute (ROCm) Runtime software stack

ii  rocminfo                   1.0.0              amd64              Radeon Open Compute (ROCm) Runtime rocminfo tool

diego@diego-MS-7B09:[master]~/Git/rocBLAS$ dpkg -l | grep hip

ii  hip_base                   1.5.19284          amd64              HIP: Heterogenous-computing Interface for Portability [BAS

ii  hip_doc                    1.5.19284          amd64              HIP: Heterogenous-computing Interface for Portability [DOC

ii  hip_hcc                    1.5.19284          amd64              HIP: Heterogenous-computing Interface for Portability [HCC

ii  hip_samples                1.5.19284          amd64              HIP: Heterogenous-computing Interface for Portability [SAM

ii  whiptail                   0.52.18-1ubuntu2   amd64              Displays user-friendly dialog boxes from shell scripts

diego@diego-MS-7B09:[master]~/Git/rocBLAS$

**Comment by zaliu** [comment]
Please use **master** branch and make sure your machine has 64GB main memory if you would like to do a full build including gfx908 support (then you wouldn't be using ROCm 2.7).  If your machine has 32GB main memory, please use the -lasm_ci flag with install.sh to reduce memory footprint.  I can assure you we've done full builds on 64GB main memory machines as well as -lasm_ci builds (which cover gfx900 and gfx906) on 32GB main memory machines successfully many, many times.

It will take about an hour on a machine with 32GB main memory to finish a -lasm_ci build.  From time to time, the build seems to be stuck.  It is not.  The compiler is actually work hard doing its job.  Run `top` to see when component of the compiler is working at any point.  I understand it's trying to wait for even a -lasm_ci build to finish.

If your machine has only 16GB main memory, please make sure you have at least 16GB swap configured, and make sure you do only -lasm_ci builds.  It's going to take about 80 minutes to finish a -lasm_ci build with such limited memory configuration.  Full builds will not succeed.

In summary, have at least 32GB main memory, use -lasm_ci flag to install.sh, and be very patient.

Please do not use develop branch.  Thank you in advance.

**Comment by xdgarrido** [comment]
I have 32GB and I did use the -lasm_ci builds. It crash the system after hours... I will upgrade 
to 64 GB and try out....

**Comment by zaliu** [comment]
@pfultz2, I just built rocBLAS master-rocm-2.7 with ROCm 2.7 just fine.  Please share details of your build environment offline.  Thanks.

**Comment by zaliu** [comment]
Just FYI, for the latest master branch (336cecd990b023a4032946c25e42477a607f62e2), the observed memory footprint was about 23GB (via `free -h` every second) when doing a install.sh -lasm_ci build with ROCm 2.7 on a 32GB machine.  It took about 54 minutes to finish the build.

**Comment by pfultz2** [comment]
> I just built rocBLAS master-rocm-2.7 with ROCm 2.7 just fine. Please share details of your build environment offline.

I build it doing something like this:

```
mkdir ~/tmp/rocblas-build
cd ~/tmp/rocblas-build
CXX=hcc cmake ~/rocBLAS
```

Or doing `cget init --cxx /opt/rocm/bin/hcc && cget install ROCmSoftwarePlatform/rocBLAS`. The problem seems to be the hardcoded paths to a build directory. I am working on a fix.

**Comment by xdgarrido** [comment]
Please, remove old rocblas library before compilation and everything should work...

**Comment by pfultz2** [comment]
PR #670 fixes this issue.
